### ISSUE TITLE
SelectorFormat doesn't expand parent references for nested selectors

### ISSUE THREAD

**Issue by mattberridge** [causes/scss-lint#339]
I've created a custom regex for a selector naming convention I am using:
`^[cul]-[A-Z]([a-z]+|_{2}|[^-]-{1,2})+$`

This is working (tested on Rubular) and it works too when linting, but only for fully written selectors in my `.scss` partials.

For example, I've disallowed single underscores. And this happens:

```
// fails when linted
.c-Media_body {
    ...
}

// doesn't fail when linted
.c-Media {
    &_body {
        ...
    }
}
```

Is there something I've done wrong here, or is this a limitation with the gem and the new Sass 3.4 Bem-style selector feature?

Thanks

**Comment by sds** [comment]
Hey @mattberridge, thanks for reporting this issue.

The problem here is the lack of expansion of the parent reference `&` in `&_body`. This is actually not entirely trivial to do, as we need to manually crawl up the rule set hierarchy through each selector and finding all possible selectors that could be generated. You can see how the number of potential selectors we'd have to check can grow quickly with this example:

```scss
.a1,
.a2,
.a3 {
  &-b1,
  &-b2,
  &-b3 {
    &-suffix {
       ...
    }
  }
}
```

The above example would generate the following selectors:

```
.a1-b1-suffix
.a1-b2-suffix
.a1-b3-suffix
.a2-b1-suffix
.a2-b2-suffix
.a2-b3-suffix
.a3-b1-suffix
.a3-b2-suffix
.a3-b3-suffix
```

To handle this in the general case is possible (assuming we ignore any interpolation), but would take some work to implement.

Would love to see a pull request for this if this is the level of detail that people want from this linter, but the juice probably isn't worth the squeeze from my perspective.

**Comment by sds** [comment]
Looking into this a little deeper, I think I need to withdraw my statement about this being possible in the general case.

Suppose you had a convention that all class names in selectors are of the form `.<word>-<word>`, i.e. every class has exactly one hyphen.

Now consider the following SCSS:

```scss
.one {
  &-two { ... }

  &-three { ... }
}
```

In this case, you would expect no lints, since the classes `.one-two` and `.one-three` are rendered. However, the linter will still report a lint for `.one`, since it technically doesn't contain a hyphen, despite the fact that it won't get rendered on its own. `scss-lint` doesn't know this though&mdash;all it sees when it visits `.one` is that it is a class selector.

I don't think it's the place of `scss-lint` to figure out all the clever ways you can render selectors with Sass and gracefully handle those cases. It could probably be made more clever for some cases, but as I mentioned before I don't think the juice is worth the squeeze.

I would think of parent references as _dynamic_ content. It takes some actual execution of the language and understanding its semantics in order to evaluate a parent reference, and `scss-lint` should be seen as a _static_ analysis tool. Obviously that line can be blurred a bit, but I don't think it's in the project's interest to cross that line in this case.
